Profile support

brockelmore opened the pull request:
Added profile support. Adds 3 `Effect`s - `ActivateProfile` `DeactivateProfile` and `DeactivateAllProfiles`.

Allows you to define a profile as having aliases or indexes explicitly enabled or disabled. This is done to avoid making users manage layers and instead focus on keybinding conflict resolution.

Let me know your thoughts on structure/what you would like changed

brockelmore committed:
profile support

ItayGarin commented on examples/cfg.ron:
```diff
@@ -23,6 +23,21 @@
         "i_is_enter": 4,
     },
 
+    layer_profiles: {
```
I'm wondering whether the "off" profile fields make these a bit verbose/awkward.
Intuitively, when I try to imagine what I'd want a profile to look like, I only think about which layers should be on.
I'd expect the rest of the layers to be off.

Would that also work with the workflow you had in mind? 
Turning off all layers, then applying the layers in the profile.

brockelmore replied:
Eh yeah you're probably right where we only have one profile at a time. Again I was thinking of having multiple profiles loaded, but we can just do one if you think that is better. I was kind of thinking something where you have a base profile for an app and you could add extra profiles. But if you think that "extra" profile should just be entirely encompassing of all functionality for the user, I think that makes sense

ItayGarin replied:
Yeah. I think it'll be simpler / more intuitive. 
Though, we can definitely iterate on this again in the future if we find my suggestion lacking.

ItayGarin commented on examples/cfg.ron:
```diff
@@ -40,42 +55,13 @@
-
-            // F4 on regular tap, layer toggle by index otherwise.
-            KEY_F4:  TapHold(Key(KEY_F4), ToggleLayerAlias("i_is_enter")),
+            KEY_F1:  TapHold(Key(KEY_F1), Multi([DeactivateProfile("not-firefox"), ActivateProfile("firefox")])),
```
With my above suggestion, I think these could be a single ActivateProfile effect.

ItayGarin commented on src/effects/perform.rs:
```diff
@@ -90,6 +90,24 @@ fn perform_toggle_layer_alias(ktrl: &mut Ktrl, name: String, value: KeyValue) ->
+    Ok(())
+}
+
+fn perform_toggle_all_profiles(ktrl: &mut Ktrl, value: KeyValue) -> Result<(), Error> {
```
I'd rename this to `perform_deactivate_all_profiles` :)

brockelmore replied:
Yep sounds good

ItayGarin commented on src/layers.rs:
```diff
@@ -281,6 +295,64 @@ impl LayersManager {
 
+    pub fn turn_alias_on(&mut self, name: String) {
+        if let Some(index) = self.get_idx_from_alias(name) {
+            let idx = index.clone();
```
nitpick: I think you don't need a clone here. usize should implement `Copy`

brockelmore replied:
You get a mutable borrow reservation conflict if you don't clone. Currently a warning, but really probably should be an error. The clone almost certainly would be optimized away by compiler if it isn't necessary

ItayGarin replied:
Haha gotcha. No worries then

ItayGarin commented on src/layers.rs:
```diff
@@ -281,6 +295,64 @@ impl LayersManager {
+    }
+
+    // used as short hand to check state prior to turning off. Essentially toggle, but explicit
+    pub fn turn_index_off(&mut self, idx: usize) {
```
I think we can safely adjust `turn_layer_off/on` to do this by default.

I think that these two new functions introduce a bit of ambiguity.
I.E if you only had access to the function names. What would you expect
`turn_layer_on` to do vs `turn_index_on`

brockelmore replied:
Yep agree - almost did that but figured I wouldn't touch it for now and could adjust later. Will change

ItayGarin reviewed (changes requested):
Thanks again for the PR @brockelmore !
Let me know what you think about my comments :)

ItayGarin commented:
FYI - I added two comments before the review. Should've included them in the review itself.
I just wanted to make sure you don't miss them. GitHub's UI can be a bit overwhelming at times haha

brockelmore committed:
ActivateProfile deactivates all other profiles

Removed unnecessary functions

Made `turn_layer_{}` with no asserts and insteads uses if statements

removed on_/off_ prefix for profiles

updated config to be better

brockelmore committed:
cargo fmt

ItayGarin commented:
LGTM
Thanks!! Merged :)
